### ISSUE TITLE
XML - Crash on missing key

### ISSUE THREAD

**Issue by Alarson93** [evermeer/EVReflection#270]
Current using the following:
```
    pod 'EVReflection/Realm', :git => 'https://github.com/evermeer/EVReflection.git', :tag => '5.4.3'
    pod 'EVReflection/XML', :git => 'https://github.com/evermeer/EVReflection.git', :tag => '5.4.3'
```

Example XML:
```
<?xml version="1.0" encoding="UTF-8"?>

<xmldata xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">

<filename>MyFileName.pdf</filename>

</xmldata>
```

Example object:
class Result: Object, EVReflectable {
```
    @objc dynamic var filename: String?
}
```

When EVR attempts to turn the XML into a Dictionary, the `<xmldata xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">` line is interpreted as a field.
Its key is "__name" and value is "xmldata". 
This occurs on line 20 of EVReflectable+XMLDictionary.

I figured this wouldn't be a big deal. I'm not including the "__name" key as a property so I assume that key-value pair will simply be discarded. But, to my surprise, EVR attempts to set the data to the appropriate (non-existant) property. This occurs in the `setPropertiesfromDictionary` function because "__name" shows up in the keyMapping list.

I dived into `getKeyMapping` to figure out why this would be. 
It would seem that you include the unknown mappings. Specifically the following:
```
if !isAdded {
    keyMapping.append((item.key as? String, item.key as? String))
}
```

We are including a key that does not have a corresponding property on a Realm object. This appears to cause a crash in RLMObjectBase.

In summary:
I am combining the Realm and XML extensions. A crash occurs when a key exists in the XML that does not have a complimentary property in the object.

**Comment by evermeer** [comment]
When using the default EVReflection you would use EVObject as the base object. EVObject has an implementation of the setValue forUndefinedKey function which wil just log a warning when an unknown key is used. 

Adding the unknown keys to the dictionary is done on purpose. There are various reasons why you should want this. For instance you could add a validation mechanism to your mapping to make sure it's still in line with your object definition. Or you could have some custom field mapping or conversion.

But... You are linking to EVReflection 5.4.3... In version 5.4.5 I added functionality to handle this exact situation. Please try that version.

If you still would like to use the 5.4.3 version... In your situation you are using the Realm Object as the base class. Realm has some extra non default handling of the setValue forKey and setValue forUndefinedKey functions. The cleanest way to handle your situation is by adding a propertyMapping for the field __name like this:

```
open func propertyMapping() -> [(keyInObject: String?, keyInResource: String?)] {
        return [(keyInObject: nil, keyInResource: "__name")]
    }
```

If you would like the behavior that your app would not crash when there is an unknown key in the xml, then you have to implement the setValue forUndefinedKey function like I did in the 5.4.5 version

**Comment by Alarson93** [comment]
Thank you for your prompt response!

Ahh the version - I suppose I should have checked that first; I've been locking my pod versions to mitigate problems on older projects.

Version 5.4.5 appears to introduce a new problem. `let` properties such as RealmList<String> will trigger value:forUndefinedKey: and, as a result of your override, will return nil. This causes problems when I make a [detached copy](https://github.com/realm/realm-cocoa/issues/5433) from Realm.

The good news is I believe I have a solution! Since this override is an extension of the Realm Object class - we can look at the schema and check if the object has the associated property.

The following is 
```
    override open func value(forUndefinedKey key: String) -> Any? {
        if objectSchema.properties.map({$0.name}).contains(key) == true {
            return super.value(forUndefinedKey: key)
        }
        
        return nil
    }
```

This appears to have worked in the case of the project I'm working on; unused keys are ignored, but special realm properties (such as RealmLists) are included.

**Comment by evermeer** [comment]
😃 Always fun when people use your code in ways you never thought about...

And awesome that you also found a workaround! 
I will have a look at this and will implement it as soon as possible (if there are no side effects)

**Comment by Alarson93** [comment]
Absolutely :)

There aren't very many reflection based libraries for Mac/iOS development.
Apple has their own reflection-based parsing now (via Codable), but I feel that it is not very flexible and it doesn't play well with Realm.
There aren't many good XML parsers for Mac/iOS development, let alone parsers that convert into a dictionary.

Your library hits on all the points I'm looking for. And - to the best of my knowledge - nothing else does.

Thank you for your efforts in supporting this unique library!

**Comment by evermeer** [comment]
I have just pushed version 5.5.4 to GitHub. The publication to Cocoapods is still running. This usually takes about ½ an hour and sometimes needs a 2nd attempt :)

**Comment by Alarson93** [comment]
Thank you! I will keep an eye out. Currently the highest version is 5.5.3 - but I assume it is still processing.

My experience with releasing and updating public Cocoapods is fairly limited - but I recall a very long wait time (~1 day?) before my first public pod was available on the main Cocoapod trunk.

**Comment by evermeer** [comment]
I have various pods. The smallest is published within a minute. But EVReflection has multiple subspecs which have various dependenciesUsually so usually it's about ½ an hour. But my laptop fell a sleep and I forgot to check. Now it's published.

--------------------------------------------------------------------------------
 🎉  Congrats

 🚀  EVReflection (5.5.4) successfully published
 📅  February 13th, 21:37
 🌎  https://cocoapods.org/pods/EVReflection
 👍  Tell your friends!
--------------------------------------------------------------------------------

**Comment by Alarson93** [comment]
I updated to version 5.5.4 but it would appear that the issue I mentioned earlier (with `let`) persists.

The reason is my proposed change was not included:
```
    override open func value(forUndefinedKey key: String) -> Any? {
        if objectSchema.properties.map({$0.name}).contains(key) == true {
            return super.value(forUndefinedKey: key)
        }
        
        return nil
    }
```

Did you intend to have this (or something like it) included in 5.5.4?

**Comment by evermeer** [comment]
Indeed it seems that the 5.5.4 release did not include the right commit. I created release 5.5.5 that does include it. See https://github.com/evermeer/EVReflection/commit/a9d232463f95a000976d89a188a80527d581d107
Again the Cocoapod publish is running...

**Comment by Alarson93** [comment]
Many thanks!

I'll check in about 30 minutes and, assuming it is available, I'll run the update and get back to you as to whether everything is working as expected.

**Comment by evermeer** [comment]
it's published...

**Comment by Alarson93** [comment]
Hello! I have updated, but I've run into a different problem - hence my delayed response. This also prevents me from properly testing the result of the update.

```
class MyTest: Object, EVReflectable {
    let row = List<Row>()
}

class Row: Object, EVReflectable {
    @objc dynamic var potato: String?
}
```

```
        let details = ["row" : [ ["potato" : "hello"], ["potato" : "goodbye"] ]] as! NSDictionary
        let object = MyTest.init(dictionary: details)
```

I am unable to initiate my object as the app crashes. 
The crash is on line 984 (`anyObject.setValue(setValue, forKey: key)`) in EVReflection. 

This issue only occurs on Realm objects (as in "Object, EVReflecatable"). Additionally - the issue seems to occur regardless of whether my "rows" property is an Array<Row> or a List<Row>.

**Comment by evermeer** [comment]
Hmmm... It seems like row is a reserved keyword. The moment you rename it your code will work. Do you need a workaround for this?

**Comment by Alarson93** [comment]
That appears to be the issue - thank you for spotting that!

I don't have control over the XML that is being sent back (in the real world scenario), so I cannot have the key "row" changed to something else.
I figured I could just use the propertyMapper:
```
class MyTest: Object, EVReflectable {
    let rows = List<Row>()
    
    func propertyMapping() -> [(keyInObject: String?, keyInResource: String?)] {
        return [(keyInObject: "row", keyInResource: "rows")]
    }
}

OR

class MyTest: Object, EVReflectable {
    let rows = List<Row>()
    
    func propertyMapping() -> [(keyInObject: String?, keyInResource: String?)] {
        return [(keyInObject: "rows", keyInResource: "row")]
    }
}

```

But, while this method fires, it doesn't seem to change anything. The key setValue is trying to use is still row. It seems to use the key of the dictionary/JSON/XML instead of the custom property key I provide.

**Comment by Alarson93** [comment]
Disregard the issue in the above post. The following workaround was successful:
```
class MyTest: Object, EVReflectable {
    let rows = List<Row>()
    
    func propertyMapping() -> [(keyInObject: String?, keyInResource: String?)] {
        return [(keyInObject: "rows", keyInResource: "row")]
    }
}
```

I am now able to take in data that contains the key "row" and assign it to a property named "rows". I will test with my full data set shortly and post back with the results.

Update: 
The test was successful! XML deserialized, no crashes, and the `let` issue is not present.
Thank you very much. I think it is safe to say that this issue is resolved.

**Comment by evermeer** [comment]
Great to hear that you have it working. 
It's good to know that it can really work when going from XML to Realm.
